momento-sdk/src/main/java/momento/sdk/SubscriptionWrapper.java

```diff
@@ -0,0 +1,207 @@
+import momento.sdk.responses.topic.TopicSubscribeResponse;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class SubscriptionWrapper implements Closeable {
+  private final Logger logger = LoggerFactory.getLogger(SubscriptionWrapper.class);
+  private final ScsTopicGrpcStubsManager grpcManager;
+  private final String cacheName;
+  private final String topicName;
+  private final SubscriptionState subscriptionState;
+  private final ISubscribeCallOptions options;
+  private StreamObserver<_SubscriptionItem> subscription;
+  private boolean subscribed = false;
+
+  public SubscriptionWrapper(
+      ScsTopicGrpcStubsManager grpcManager,
+      String cacheName,
+      String topicName,
+      SubscriptionState subscriptionState,
+      ISubscribeCallOptions options) {
+    this.grpcManager = grpcManager;
+    this.cacheName = cacheName;
+    this.topicName = topicName;
+    this.subscriptionState = subscriptionState;
+    this.options = options;
+  }
+
+  public CompletableFuture<Void> subscribe() {
+    CompletableFuture<Void> future = new CompletableFuture<>();
+
+    subscription =
+        new StreamObserver<_SubscriptionItem>() {
+          boolean firstMessage = true;
+
+          @Override
+          public void onNext(_SubscriptionItem value) {
+            if (firstMessage) {
+              firstMessage = false;
+              future.complete(null);
+            }
+            handleSubscriptionItem(value);
+          }
+
+          @Override
+          public void onError(Throwable t) {
+            if (firstMessage) {
+              firstMessage = false;
+              future.completeExceptionally(t);
+            } else {
+              handleSubscriptionError(t);
+            }
+          }
+
+          @Override
+          public void onCompleted() {
+            handleSubscriptionCompleted();
+          }
+        };
+
+    _SubscriptionRequest subscriptionRequest =
+        _SubscriptionRequest.newBuilder()
+            .setCacheName(cacheName)
+            .setTopic(topicName)
+            .setResumeAtTopicSequenceNumber(subscriptionState.getResumeAtTopicSequenceNumber())
+            .build();
+
+    try {
+      grpcManager.getStub().subscribe(subscriptionRequest, subscription);
+    } catch (Exception e) {
+      future.completeExceptionally(
+          new TopicSubscribeResponse.Error(CacheServiceExceptionMapper.convert(e)));
+    }
+    return future;
+  }
+
+  private void handleSubscriptionError(Throwable t) {
+    // Handle the subscription error
+    logger.info("error " + cacheName + " " + topicName + " " + t.getMessage());
+    this.options.onError(t);
+  }
+
+  private void handleSubscriptionCompleted() {
+    // Handle the subscription completion
+    logger.info("completed " + cacheName + " " + topicName);
+    this.options.onCompleted();
+  }
+
+  private void handleSubscriptionItem(_SubscriptionItem item) {
+    // Handle subscription item
+    if (!subscribed) {
```

Comment:
its a bit strange to me to have both a `firstMessage` boolean, as well as a `subscribed` boolean. I would try and consolidate the two. At first glance, I would probably try moving this logic checking for the heartbeat above where we complete the original future. If the first message is not a heartbeat, we should complete the future exceptionally